modules/host/update.go

```diff
@@ -97,23 +101,36 @@ func (h *Host) initConsensusSubscription() error {
 		// structure.
 		return h.initRescan()
 	}
-	return err
+	if err != nil {
+		return err
+	}
+	h.tg.OnStop(func() {
+		h.cs.Unsubscribe(h)
+	})
+	return nil
 }
 
 // ProcessConsensusChange will be called by the consensus set every time there
 // is a change to the blockchain.
 func (h *Host) ProcessConsensusChange(cc modules.ConsensusChange) {
-	h.mu.Lock()
-	defer h.mu.Unlock()
 	err := h.tg.Add()
 	if err != nil {
 		return
 	}
-	defer h.tg.Done()
+	wg := new(sync.WaitGroup)
+	defer func() {
+		wg.Wait()
```

Comment:
this is far away from where the waitgroup is actually used. I would move it down to the `for` loop and call `wg.Wait` directly instead of in the `defer`.